parlai/agents/torch_agent/torch_agent.py

```diff
@@ -0,0 +1,218 @@
+            label_type = 'labels'
+        elif 'eval_labels' in obs:
+            label_type = 'eval_labels'
+
+        if label_type is not None:
+            new_labels = []
+            for label in obs[label_type]:
+                vec_label = self.dict.txt2vec(label)
+                if addEndIdx:
+                    vec_label.append(self.END_IDX)
+                new_label = torch.LongTensor(vec_label)
+                if self.use_cuda:
+                    new_label = new_label.cuda()
+                new_labels.append(new_label)
+            obs[label_type] = new_labels
+
+        return obs
+
+    def permute(self, obs_batch, sort=True, is_valid=lambda obs: 'text' in obs):
+        """Creates a batch of valid observations from an unchecked batch.
+        Assumes each observation has been vectorized by vectorize function.
+        """
+        try:
+            # valid examples and their indices
+            valid_inds, exs = zip(*[(i, ex) for i, ex in
+                                    enumerate(obs_batch) if is_valid(ex)])
+        except ValueError:
+            # zero examples to process in this batch, so zip failed to unpack
+            return None, None, None, None, None, None
+        x_text = [ex['text'] for ex in exs]
+        x_lens = [ex.shape[0] for ex in x_text]
+
+        if sorted:
+            ind_sorted = sorted(range(len(x_lens)), key=lambda k: -x_lens[k])
+
+            exs = [exs[k] for k in ind_sorted]
+            valid_inds = [valid_inds[k] for k in ind_sorted]
+            x_text = [x_text[k] for k in ind_sorted]
+            end_idxs = [x_lens[k] for k in ind_sorted]
+
+        padded_xs = torch.LongTensor(len(exs),
+                                     max(x_lens)).fill_(self.NULL_IDX)
+        if self.use_cuda:
+            padded_xs = padded_xs.cuda()
+
+        for i, ex in enumerate(x_text):
+            padded_xs[i, :ex.shape[0]] = ex
+
+        xs = padded_xs
+
+        eval_labels_avail = any(['eval_labels' in ex for ex in exs])
+        labels_avail = any(['labels' in ex for ex in exs])
+        some_labels_avail = eval_labels_avail or labels_avail
+
+        # set up the target tensors
+        ys = None
+        labels = None
+        y_lens = None
+        if some_labels_avail:
+            # randomly select one of the labels to update on (if multiple)
+            if labels_avail:
+                labels = [random.choice(ex.get('labels', [''])) for ex in exs]
+            else:
+                labels = [random.choice(ex.get('eval_labels', [''])) for ex in exs]
+            y_lens = [y.shape[0] for y in labels]
+            padded_ys = torch.LongTensor(len(exs),
+                                         max(y_lens)).fill_(self.NULL_IDX)
+            if self.use_cuda:
+                padded_ys = padded_ys.cuda()
+            for i, y in enumerate(labels):
+                padded_ys[i, :y.shape[0]] = y
+            ys = padded_ys
+        return xs, ys, labels, valid_inds, end_idxs, y_lens
```

Comment:
So I looked through the repo and it seems none of our models uses that output from `pad_text`, though LanguageModel does use `y_lens`. I'll remove it from this function since it isn't used.